README.md
# Money

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Money is a module for obtaining currency rate and stock market information.

## Installation

```bash
npm install money-js
```

## Usage

```javascript
var money = require('money-js');

money.quote({symbols: [
  'USDCAD',
  'USDEUR',
  'JPYUSD',
  'TWTR',
  'NASDAQ'
]}, function(err, result) {
  if(err) console.log(err);

  console.log(result);
});
/*
[ { symbol: 'USDCAD',
    type: 'Currency',
    name: 'United States Dollar - Canadian Dollar',
    high: '1.0869',
    low: '1.0859',
    last: '1.0863',
    volume: '0',
    change: '-0.0004',
    percentchange: '-0.0368%',
    timeoflastsale: '',
    currency: '',
    ask: '1.0865',
    bid: '1.0861',
    yearhigh: '0',
    yearlow: '0',
    marketcap: '0',
    pe: '0',
    close: '1.0867',
    open: '1.0867',
    eps: '0',
    exchange: 'FOREX cross rates',
    sizeoflastsale: '0' },
  { symbol: 'USDEUR',
    type: 'Currency',
    name: 'United States Dollar - Euro',
    high: '0.7389',
    low: '0.7382',
    last: '0.7383',
    volume: '0',
    change: '-0.0005',
    percentchange: '-0.0616%',
    timeoflastsale: '',
    currency: '',
    ask: '0.7383',
    bid: '0.7382',
    yearhigh: '0',
    yearlow: '0',
    marketcap: '0',
    pe: '0',
    close: '0.7387',
    open: '0.7387',
    eps: '0',
    exchange: 'FOREX cross rates',
    sizeoflastsale: '0' },
  { symbol: 'JPYUSD',
    type: 'Currency',
    name: 'Japan Yen - United States Dollar',
    high: '0.0098',
    low: '0.0098',
    last: '0.0098',
    volume: '0',
    change: '0',
    percentchange: '-0.0837%',
    timeoflastsale: '',
    currency: '',
    ask: '0.0098',
    bid: '0.0098',
    yearhigh: '0',
    yearlow: '0',
    marketcap: '0',
    pe: '0',
    close: '0.0098',
    open: '0.0098',
    eps: '0',
    exchange: 'FOREX cross rates',
    sizeoflastsale: '0' },
  { symbol: 'TWTR',
    type: 'Equity',
    name: 'TWITTER Inc',
    high: '35.97',
    low: '34.93',
    last: '35.54',
    volume: '30,021,271',
    change: '+0.17',
    percentchange: '+0.4806%',
    timeoflastsale: '6/11/2014 7:58:48 PM',
    currency: 'USD',
    ask: '52',
    bid: '0',
    yearhigh: '74.73',
    yearlow: '29.51',
    marketcap: '20,966,425,876',
    pe: '-14.2129',
    close: '35.37',
    open: '35.07',
    eps: '-2.5006',
    exchange: 'New York Stock Exchange',
    sizeoflastsale: '0' },
  { symbol: 'NASDAQ',
    type: 'Index',
    name: 'NASDAQ Composite',
    high: '4,338.2108',
    low: '4,315.487',
    last: '4331.932',
    volume: '0',
    change: '-6.0653',
    percentchange: '-0.1398%',
    timeoflastsale: '6/11/2014 5:15:59 PM',
    currency: 'USD',
    ask: '0',
    bid: '0',
    yearhigh: '4,371.706',
    yearlow: '3,294.9495',
    marketcap: '0',
    pe: '0',
    close: '4,337.9973',
    open: '4,322.9221',
    eps: '0',
    exchange: 'NASDAQ Indices',
    sizeoflastsale: '0' } ]
*/
```

## Notes

* It uses `money.service.msn.com`

## License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/money-js
[npm-image]: https://badge.fury.io/js/money-js.svg

[travis-url]: https://travis-ci.org/devfacet/money
[travis-image]: https://travis-ci.org/devfacet/money.svg?branch=master


CHANGELOG.md
# Changelog

## 1.0.2 (2017-03-05)

- Maintenance

## 1.0.1 (2015-03-15)

- Misc. updates

## 1.0.0 (2014-07-04)

- Stable release

## 0.1.1 (2014-06-14)

- Changes for timeout

## 0.1.0 (2014-06-11)

- Initial release

test/test-all.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var money  = require('../'),
    expect = require('chai').expect;

// Tests

describe('money', function() {

  // quote
  describe('quote()', function() {
    it.skip('should get information', function(done) {
      money.quote({symbols: ['USDCAD', 'USDEUR', 'JPYUSD', 'TWTR','NASDAQ']}, function(err, result) {
        if(err) return done(err);

        expect(err).to.be.equal(undefined);

        expect(result).to.be.a('array');
        expect(result).to.have.property('length').to.be.equal(5);

        expect(result[0]).to.be.a('object');
        expect(result[0]).to.have.property('symbol', 'USDCAD');
        expect(result[0]).to.have.property('type', 'Currency');
        expect(result[0]).to.have.property('name');
        expect(result[0]).to.have.property('high');
        expect(result[0]).to.have.property('low');
        expect(result[0]).to.have.property('last');
        expect(result[0]).to.have.property('volume');
        expect(result[0]).to.have.property('change');
        expect(result[0]).to.have.property('percentchange');
        expect(result[0]).to.have.property('timeoflastsale');
        expect(result[0]).to.have.property('currency');
        expect(result[0]).to.have.property('ask');
        expect(result[0]).to.have.property('bid');
        expect(result[0]).to.have.property('yearhigh');
        expect(result[0]).to.have.property('yearlow');
        expect(result[0]).to.have.property('marketcap');
        expect(result[0]).to.have.property('pe');
        expect(result[0]).to.have.property('close');
        expect(result[0]).to.have.property('open');
        expect(result[0]).to.have.property('eps');
        expect(result[0]).to.have.property('exchange');
        expect(result[0]).to.have.property('sizeoflastsale');

        expect(result[3]).to.be.a('object');
        expect(result[3]).to.have.property('symbol', 'TWTR');
        expect(result[3]).to.have.property('type', 'Equity');

        expect(result[4]).to.be.a('object');
        expect(result[4]).to.have.property('symbol', 'NASDAQ');
        expect(result[4]).to.have.property('type', 'Index');

        done();
      });
    });
  });

});

index.js
/*
 * Money
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true, sub: true */
'use strict';

var request = require('request'),
    qs      = require('querystring'),
    xml2JS  = require('xml2js');

// Init the module
module.exports = function() {

  var xmlParser  = new xml2JS.Parser({charkey: 'C$', attrkey: 'A$', explicitArray: true}),
      defTimeout = 10000,
      quoteUrl   = 'http://money.service.msn.com/StockQuotes.aspx?v=1',
      quote;     // gets information - function

  quote = function quote(options, callback) {

    if(typeof callback !== 'function')
      callback = function callback(err, result) { return err || result; };

    if(!options || typeof options !== 'object')
      return callback('Invalid options!');

    if(!options.symbols)
      return callback('Missing symbols input!');

    if(typeof options.symbols === 'string')
      options.symbols = [options.symbols];

    if(!(options.symbols instanceof Array))
      return callback('Invalid symbols input!');

    var result,
        reqUrl  = quoteUrl + '&symbols=' + qs.escape(options.symbols.join(','));

    request.get({url: reqUrl, timeout: defTimeout}, function(err, res, body) {

      if(err) return callback(err);
      if(res.statusCode !== 200) return callback('Request failed (' + res.statusCode + ')');

      xmlParser.parseString(body, function(err, resultJSON) {
        if(err) return callback(err);

        if(!resultJSON.quotesdata || !resultJSON.quotesdata.ticker)
          return callback('Unexpected error! Invalid content.');

        if(resultJSON.quotesdata.ticker['A$'] && resultJSON.quotesdata.ticker['A$'].errormessage)
          return callback(resultJSON.quotesdata.weather['A$'].errormessage);

        if(!(resultJSON.quotesdata.ticker instanceof Array)) {
          return callback('Unexpected error! Missing info.');
        }

        for(var i = 0, tickerLen = resultJSON.quotesdata.ticker.length; i < tickerLen; i++) {

          if(typeof resultJSON.quotesdata.ticker[i]['A$'] !== 'object')
            continue;

          if(!result) result = [];
          result.push(resultJSON.quotesdata.ticker[i]['A$']);
        }

        return callback(undefined, result);
      });
    });
  };

  // Return
  return {
    quote: quote
  };
}();

lib/money.js
/*
 * Money
 * Copyright (c) 2014 Fatih Cetinkaya (http://github.com/cmfatih/money)
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var request = require('request'),
    qs      = require('querystring'),
    xml2JS  = require('xml2js');

// Init the module
exports = module.exports = function() {

  var xmlParser  = new xml2JS.Parser({charkey: 'C$', attrkey: 'A$', explicitArray: true}),
      defTimeout = 10000,
      quoteUrl   = 'http://money.service.msn.com/StockQuotes.aspx?v=1',
      quote;     // gets information - function

  quote = function quote(options, callback) {

    if(typeof callback !== 'function')
      callback = function callback(err, result) { return err || result; };

    if(!options || typeof options !== 'object')
      return callback('Invalid options!');

    if(!options.symbols)
      return callback('Missing symbols input!');

    if(typeof options.symbols === 'string')
      options.symbols = [options.symbols];

    if(!(options.symbols instanceof Array))
      return callback('Invalid symbols input!');

    var result,
        reqUrl  = quoteUrl + '&symbols=' + qs.escape(options.symbols.join(','));

    request.get({url: reqUrl, timeout: defTimeout}, function(err, res, body) {

      if(err) return callback(err);
      if(res.statusCode !== 200) return callback('Request failed (' + res.statusCode + ')');

      xmlParser.parseString(body, function(err, resultJSON) {
        if(err) return callback(err);

        if(!resultJSON.quotesdata || !resultJSON.quotesdata.ticker)
          return callback('Unexpected error! Invalid content.');

        if(resultJSON.quotesdata.ticker['A$'] && resultJSON.quotesdata.ticker['A$'].errormessage)
          return callback(resultJSON.quotesdata.weather['A$'].errormessage);

        if(!(resultJSON.quotesdata.ticker instanceof Array)) {
          return callback('Unexpected error! Missing info.');
        }

        for(var i = 0, tickerLen = resultJSON.quotesdata.ticker.length; i < tickerLen; i++) {

          if(typeof resultJSON.quotesdata.ticker[i]['A$'] !== 'object')
            continue;

          if(!result) result = [];
          result.push(resultJSON.quotesdata.ticker[i]['A$']);
        }

        return callback(undefined, result);
      });
    });
  };

  // Return
  return {
    quote: quote
  };
}();